sw/device/tests/sram_ctrl_execution_test_main.c

```diff
@@ -0,0 +1,163 @@
+     *   +----------------+-------+-------+------+-----------+
+     *   31             20 19   15 14   12 11   7 6          0
+     *
+     *   Custom representation: jalr zero, 0(ra)
+     *   +----------------+---------+-------+---------+-----------+
+     *   |  000000000000  |  00001  |  000  |  00000  |  1100111  |
+     *   +----------------+---------+-------+---------+-----------+
+     *   31             20 19     15 14   12 11      7 6          0
+     */
+    0x00008067,
+};
+
+/*
+ * Adds the first parameter to the second parameter and returns the result.
+ *
+ * Maps to the `execution_test_instructions` buffer as the function body.
+ */
+typedef uint32_t (*ram_exec_function_t)(volatile uint32_t, volatile uint32_t);
+
+static bool otp_ifetch_enabled(void) {
+  otp_ctrl_t otp = otp_ctrl_init();
+
+  dif_otp_ctrl_config_t config = {
+      .check_timeout = 100000,
+      .integrity_period_mask = 0x3ffff,
+      .consistency_period_mask = 0x3ffffff,
+  };
+  CHECK_DIF_OK(dif_otp_ctrl_configure(&otp.dif, config));
+
+  CHECK_DIF_OK(dif_otp_ctrl_dai_read_start(&otp.dif, kDifOtpCtrlPartitionHwCfg,
+                                           kOtpIfetchHwRelativeOffset));
+
+  otp_wait_for_dai(&otp);
+
+  uint32_t value;
+  CHECK_DIF_OK(dif_otp_ctrl_dai_read32_end(&otp.dif, &value));
+
+  // OTP stores IFETCH state in a single bit (enabled/disabled).
+  return value & 0x00000001;
+}
+
+/*
+ * Performs SRAM execution test.
+ *
+ * `ram_exec_function_t` is "mapped" onto the `execution_test_instructions`
+ * buffer that holds hand-crafted machine instructions. Effectively, this
+ * buffer becomes the function body.
+ *
+ * According to the calling convention the parameters will be passed in the
+ * a0 and a1 registers, with a0 also treated as the return value register.
+ */
+static void sram_execution_test(void) {
+  // Map the function pointer onto the instruction buffer.
+  ram_exec_function_t func = (ram_exec_function_t)execution_test_instructions;
+
+  LOG_INFO("Executing from SRAM at = %x", func);
+  LOG_INFO("Instruction[0] = %x (add a0, a1, a0), at %x",
+           execution_test_instructions[0], &execution_test_instructions[0]);
+  LOG_INFO("Instruction[1] = %x (jalr ra, x0) at %x",
+           execution_test_instructions[1], &execution_test_instructions[1]);
+
+  uint32_t expected_result = kExecParam1 + kExecParam2;
+  uint32_t result = func(kExecParam1, kExecParam2);
+  CHECK(result == expected_result,
+        "SRAM Execution expected %d + %d = %d, got %d", kExecParam1,
+        kExecParam2, expected_result, result);
+}
+
+/*
+ * Performs the tests.
+ *
+ * When chip is in one of the lifecycle states where debug functions are
+ * enabled, execution form SRAM is enabled providing the EN_SRAM_IFETCH
+ * (OTP) is disabled. When EN_SRAM_IFETCH (OTP) is enabled, EXEC CSR
+ * determines whether the execution from SRAM is enabled.
+ */
+bool test_main(void) {
+  lc_ctrl_t lc = lc_ctrl_init();
```

Comment:
I think we can use the DIF directly here.

Reply:
Done.